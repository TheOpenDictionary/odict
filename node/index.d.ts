/* auto-generated by NAPI-RS */
/* eslint-disable */
export declare class Dictionary {
  constructor(
    pathOrAlias: string,
    options?: DictionaryOptions | undefined | null,
  );
  static write(
    xmlStr: string,
    outPath: string,
    options?: DictionaryOptions | undefined | null,
  ): Dictionary;
  static compile(
    xmlPath: string,
    outPath?: string | undefined | null,
    options?: DictionaryOptions | undefined | null,
  ): Dictionary;
  get path(): string;
  lookup(
    query: LookupQuery | string | Array<LookupQuery | string>,
    options?: LookupOptions | undefined | null,
  ): Array<Array<Entry>>;
  lexicon(): Array<string>;
  split(query: string, options?: SplitOptions | undefined | null): Array<Entry>;
  index(options?: IndexOptions | undefined | null): void;
  search(
    query: string,
    options?: SearchOptions | undefined | null,
  ): Array<Entry>;
}

export interface Definition {
  id?: string;
  value: string;
  examples: Array<Example>;
  notes: Array<Note>;
}

export interface DictionaryOptions {
  split?: SplitOptions;
  index?: IndexOptions;
  search?: SearchOptions;
}

export interface Entry {
  term: string;
  seeAlso?: string;
  etymologies: Array<Etymology>;
}

export interface Etymology {
  id?: string;
  pronunciation?: string;
  description?: string;
  senses: Record<string, Sense>;
}

export interface Example {
  value: string;
}

export interface Group {
  id?: string;
  description: string;
  definitions: Array<Definition>;
}

export interface IndexOptions {
  directory?: string;
  memory?: number;
  overwrite?: boolean;
}

export interface LookupOptions {
  split?: number;
  follow?: boolean;
}

export interface LookupQuery {
  term: string;
  fallback: string;
}

export interface Note {
  id?: string;
  value: string;
  examples: Array<Example>;
}

export interface SearchOptions {
  directory?: string;
  threshold?: number;
  autoindex?: boolean;
  limit?: number;
}

export interface Sense {
  pos: string;
  definitions: Array<Definition | Group>;
}

export interface SplitOptions {
  threshold?: number;
}
