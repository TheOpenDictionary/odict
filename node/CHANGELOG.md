# Changelog





## [1.0.0](https://github.com/TheOpenDictionary/odict/compare/node-v0.0.1...node-v1.0.0) (2024-12-25)


### Features

* Add stable Python binding ðŸŽ‰ ([#1049](https://github.com/TheOpenDictionary/odict/issues/1049)) ([73c6e33](https://github.com/TheOpenDictionary/odict/commit/73c6e339b8614c6eb048de4ee7586dd5aa98803e))
* **ci:** Add workflow dispatch trigger for Node release ([24ce45b](https://github.com/TheOpenDictionary/odict/commit/24ce45b8678edcf92779031104c9b21614318bfa))
* Deployment changes ([#1002](https://github.com/TheOpenDictionary/odict/issues/1002)) ([981c223](https://github.com/TheOpenDictionary/odict/commit/981c2232fe8908cb9a0afd95f6c04e32a4c698ed))
* Rename Python library ([928343a](https://github.com/TheOpenDictionary/odict/commit/928343a7df53d64aa25d7e262f21f4aa0f09cc5e))
* Try updating CI ([c76c684](https://github.com/TheOpenDictionary/odict/commit/c76c684d427ef79df81e4c9a349dadba3a54339c))
* **v2:** Add Rust Node bindings ([#760](https://github.com/TheOpenDictionary/odict/issues/760)) ([aac5501](https://github.com/TheOpenDictionary/odict/commit/aac550181f6d144649ce9ad0ff823967b29668bf))


### Bug Fixes

* **ci:** Remove charabia from Node package ([cf02d0c](https://github.com/TheOpenDictionary/odict/commit/cf02d0c6f7a6b9015c88e563e7d24a846428b145))
* **ci:** Test JS CI ([86adb85](https://github.com/TheOpenDictionary/odict/commit/86adb853bd772f94d191c38406ed9e3e7b78ba28))
* **ci:** Test JS CI ([c6230cd](https://github.com/TheOpenDictionary/odict/commit/c6230cde4137ff578839380ef8b18b34f0c226ac))

## 1.0.0 (2024-12-24)


### Features

* Add stable Python binding ðŸŽ‰ ([#1049](https://github.com/TheOpenDictionary/odict/issues/1049)) ([73c6e33](https://github.com/TheOpenDictionary/odict/commit/73c6e339b8614c6eb048de4ee7586dd5aa98803e))
* **ci:** Add workflow dispatch trigger for Node release ([24ce45b](https://github.com/TheOpenDictionary/odict/commit/24ce45b8678edcf92779031104c9b21614318bfa))
* Deployment changes ([#1002](https://github.com/TheOpenDictionary/odict/issues/1002)) ([981c223](https://github.com/TheOpenDictionary/odict/commit/981c2232fe8908cb9a0afd95f6c04e32a4c698ed))
* Rename Python library ([928343a](https://github.com/TheOpenDictionary/odict/commit/928343a7df53d64aa25d7e262f21f4aa0f09cc5e))
* Try updating CI ([c76c684](https://github.com/TheOpenDictionary/odict/commit/c76c684d427ef79df81e4c9a349dadba3a54339c))
* **v2:** Add Rust Node bindings ([#760](https://github.com/TheOpenDictionary/odict/issues/760)) ([aac5501](https://github.com/TheOpenDictionary/odict/commit/aac550181f6d144649ce9ad0ff823967b29668bf))


### Bug Fixes

* **ci:** Remove charabia from Node package ([cf02d0c](https://github.com/TheOpenDictionary/odict/commit/cf02d0c6f7a6b9015c88e563e7d24a846428b145))
* **ci:** Test JS CI ([86adb85](https://github.com/TheOpenDictionary/odict/commit/86adb853bd772f94d191c38406ed9e3e7b78ba28))
* **ci:** Test JS CI ([c6230cd](https://github.com/TheOpenDictionary/odict/commit/c6230cde4137ff578839380ef8b18b34f0c226ac))
