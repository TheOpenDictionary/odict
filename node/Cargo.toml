[package]
edition = "2021"
name = "odict_node"
version = "1.1.1"
publish = false

[lib]
crate-type = ["cdylib"]

[features]
default = []
node = ["odict/http", "internal/load", "odict/alias", "odict/search"]

[dependencies]
merge = "0.2.0"
# Default enable napi4 feature, see https://nodejs.org/api/n-api.html#node-api-version-matrix
napi = { version = "3", default-features = false, features = [
  "async",
  "napi4",
] }
napi-derive = "3"
odict = { path = "../lib", features = [
  "tokenize-latin",
  "tokenize-chinese",
  "tokenize-thai",
  "tokenize-khmer",
  "tokenize-swedish",
  "tokenize-german",
  "sql",
] }
structural-convert = { workspace = true }
internal = { path = "../internal" }

[build-dependencies]
napi-build = "2"
