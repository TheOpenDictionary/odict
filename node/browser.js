export * from '@odict/node-wasm32-wasi'
