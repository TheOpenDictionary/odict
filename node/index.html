<html>
  <head>
    <title>Demo</title>
  </head>
  <body>
    <button class="pick-file">Pick File</button>
    <script type="module">
      import { Buffer } from "buffer";

      global.Buffer = Buffer;

      import { compile, Dictionary } from "./node.wasi-browser.js";

      // console.log(compile("<xml></xml>"));
      let fileHandles;

      const options = {
        types: [
          {
            description: "ODict Files",
            accept: {
              "text/binary": ".odict",
            },
          },
        ],
        excludeAcceptAllOption: true,
      };

      document.querySelector(".pick-file").onclick = async () => {
        fileHandles = await window.showOpenFilePicker(options);

        const allContent = await Promise.all(
          fileHandles.map(async (fileHandle) => {
            const file = await fileHandle.getFile();
            const content = await file.arrayBuffer();
            return content;
          }),
        );
        const dictionary = new Dictionary(new Buffer(allContent[0]), {});

        console.log(allContent[0]);

        console.log(dictionary);
        // console.log(
        //   await Dictionary.compile(
        //     '<dictionary><entry term="dog"/></entry></dictionary>'
        //   )
        // );
        console.log(dictionary.lookup("dog"));
        // console.log(dictionary.lexicon());
      };
    </script>
  </body>
</html>
