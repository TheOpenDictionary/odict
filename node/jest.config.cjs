module.exports = {
	testEnvironment: "node",
	clearMocks: true,
	extensionsToTreatAsEsm: [".ts"],
	collectCoverage: true,
	coverageDirectory: "coverage",
	coverageProvider: "v8",
	transform: {
		"^.+\\.(t|j)sx?$": ["@swc/jest"],
	},
	moduleNameMapper: {
		"^(\\.{1,2}/.*)\\.js$": "$1",
	},
};
