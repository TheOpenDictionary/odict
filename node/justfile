@deps: install
  
@install:
  pnpm install --no-frozen-lockfile

@add +args:
  pnpm add {{args}}

@npm +args:
  pnpm npm {{args}}

@clean:
  rm -rf coverage ../**/*.odict  ../**/*.log

@build:
  pnpm build 

@universal: 
  pnpm universal

@version:  
  pnpm version

@test: build && clean
  pnpm test

@publish +args="":
	just npm publish {{args}}