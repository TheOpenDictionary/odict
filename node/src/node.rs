use napi::bindgen_prelude::*;

use odict::ToDictionary;

use crate::types::LoadOptions;
use crate::types::SaveOptions;
use crate::{
  types::{self, Entry},
  utils::cast_error,
};

#[napi]
pub struct OpenDictionary {
  dict: odict::OpenDictionary,
}

#[napi]
pub fn compile(xml: String) -> Result<Buffer> {
  let bytes = xml
    .to_dictionary()
    .and_then(|d| d.build())
    .and_then(|d| d.to_bytes())
    .map_err(cast_error)?;
  Ok(bytes.into())
}

#[napi]
impl OpenDictionary {
  #[napi(constructor)]
  pub fn new(data: Buffer) -> Result<Self> {
    let dict = crate::shared::new_from_bytes(&data)?;
    Ok(Self { dict })
  }

  #[napi(getter)]
  pub fn min_rank(&self) -> Result<Option<u32>> {
    crate::shared::get_min_rank(&self.dict)
  }

  #[napi(getter)]
  pub fn max_rank(&self) -> Result<Option<u32>> {
    crate::shared::get_max_rank(&self.dict)
  }

  #[napi(factory)]
  pub async fn load(dictionary: String, options: Option<LoadOptions>) -> Result<Self> {
    use internal::LoadDictionaryOptions;

    let opts = options
      .map(|o| LoadDictionaryOptions::try_from(o))
      .transpose()
      .map_err(cast_error)?
      .unwrap_or_default();

    let dict = internal::load_dictionary_with_options(&dictionary, opts)
      .await
      .map_err(cast_error)?;

    Ok(Self { dict })
  }

  #[napi]
  pub fn save(&mut self, path: String, options: Option<SaveOptions>) -> Result<()> {
    match options {
      Some(opts) => {
        let compiler_options = odict::compile::CompilerOptions::from(opts);
        self
          .dict
          .to_disk_with_options(&path, compiler_options)
          .map_err(cast_error)
      }
      None => self.dict.to_disk(&path).map_err(cast_error),
    }
  }

  #[napi]
  pub fn lookup(
    &self,
    query: Either<String, Vec<String>>,
    options: Option<types::LookupOptions>,
  ) -> Result<Vec<types::LookupResult>> {
    crate::shared::perform_lookup(&self.dict, query, options)
  }

  #[napi]
  pub fn lexicon(&self) -> Result<Vec<&str>> {
    crate::shared::get_lexicon(&self.dict)
  }

  #[napi]
  pub fn index(&self, options: Option<types::IndexOptions>) -> Result<()> {
    let dict = self.dict.contents().map_err(cast_error)?;
    let opts = options.unwrap_or(types::IndexOptions::default());

    dict
      .index(odict::index::IndexOptions::from(opts))
      .map_err(cast_error)?;

    Ok(())
  }

  #[napi]
  pub fn search(&self, query: String, options: Option<types::SearchOptions>) -> Result<Vec<Entry>> {
    let dict = self.dict.contents().map_err(cast_error)?;
    let opts = options.unwrap_or_default();

    // Use our helper function to avoid orphan rule issues
    let results = dict
      .search(query.as_str(), odict::search::SearchOptions::from(opts))
      .map_err(cast_error)?;

    // Use the new from_entry function for Entry types
    let entries = results
      .iter()
      .map(|e| e.clone().into())
      .collect::<Vec<Entry>>();

    Ok(entries)
  }

  #[napi]
  pub fn tokenize(
    &self,
    text: String,
    options: Option<types::TokenizeOptions>,
  ) -> Result<Vec<types::Token>> {
    crate::shared::perform_tokenization(&self.dict, &text, options)
  }
}
