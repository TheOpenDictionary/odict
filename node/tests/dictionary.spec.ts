import { beforeAll, describe, expect, it } from "vitest";

import { existsSync } from "node:fs";
import { rm, stat } from "node:fs/promises";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

import { Dictionary } from "../index.js";

describe("Dictionary", () => {
  expect.addSnapshotSerializer({
    test: (t) => typeof t.value === "string",
    serialize: (t) => `"${t.value}"`,
  });

  let dict1Path: string;
  let dict2Path: string;
  let dict1: Dictionary;
  let dict2: Dictionary;

  beforeAll(async () => {
    dict1Path = join(
      fileURLToPath(new URL(import.meta.url)),
      "../../../examples/example1.xml"
    );

    dict1 = await Dictionary.compile(dict1Path);

    dict2Path = join(
      fileURLToPath(new URL(import.meta.url)),
      "../../../examples/example2.xml"
    );

    dict2 = await Dictionary.compile(dict2Path);

    const stat1 = await stat(dict1.path);
    const stat2 = await stat(dict2.path);

    expect(stat1.isFile).toBeTruthy();
    expect(stat2.isFile).toBeTruthy();
  });

  it("returns the path correctly", () => {
    // Needs toContain instead of toBe seeing Windows paths start with \\?\
    expect(dict1.path).toContain(dict1Path.replace(".xml", ".odict"));
    expect(dict2.path).toContain(dict2Path.replace(".xml", ".odict"));
  });

  describe("lookup", () => {
    it("looks up terms properly", async () => {
      const result = await dict1.lookup({ term: "cat", fallback: "cat" });
      expect(result).toMatchSnapshot();
    });

    it("doesn't split unless specified", async () => {
      const result = await dict1.lookup("catdog");
      expect(result[0].length).toBe(0);
    });

    it("can split terms", async () => {
      const result = await dict1.lookup("catdog", { split: 3 });
      expect(result).toMatchSnapshot();
    });
  });

  it("can return the lexicon", async () => {
    const result = await dict1.lexicon();
    expect(result).toStrictEqual(["cat", "dog", "poo", "ran", "run"]);
  });

  it("can write raw XML", async () => {
    await Dictionary.write(
      '<dictionary><entry term="hello"><ety><sense pos="v"><definition value="hello world" /></sense></ety></entry><entry term="world"><ety><sense pos="v"><definition value="hello world" /></sense></ety></entry></dictionary>"    )',
      "test.odict"
    );

    expect(existsSync("test.odict")).toBeTruthy();

    const dict = new Dictionary("test.odict");

    expect(dict.lookup("hello").length).toBe(1);

    await rm("test.odict");
  });

  it("can split terms properly", async () => {
    const result = await dict1.split("catdog", { threshold: 2 });
    expect(result).toMatchSnapshot();
  });

  it("can index and search a dictionary", async () => {
    await dict1.index();

    const results = await dict1.search("run");

    expect(results).toMatchSnapshot();
  });

  it("throws errors inside JavaScript", async () => {
    try {
      const dict = new Dictionary("fake-alias");
      await dict.lookup("dog");
    } catch (e) {
      expect((e as Error).message).toContain(
        "(os error 2)"
      );
    }
  });
});
