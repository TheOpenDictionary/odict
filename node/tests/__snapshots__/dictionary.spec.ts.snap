// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dictionary > can index and search a dictionary 1`] = `
[
  {
    "etymologies": [
      {
        "description": "Latin root",
        "senses": {
          "n": {
            "definitions": [
              {
                "examples": [
                  {
                    "value": "The dog runs after the cat.",
                  },
                  {
                    "value": "The horse ran away.",
                  },
                ],
                "notes": [],
                "value": "(vertebrates) To move swiftly",
              },
              {
                "examples": [],
                "notes": [],
                "value": "Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip.",
              },
              {
                "examples": [],
                "notes": [],
                "value": "A pleasure trip.",
              },
              {
                "examples": [],
                "notes": [],
                "value": "Flight, instance or period of fleeing.",
              },
              {
                "examples": [],
                "notes": [],
                "value": "Migration (of fish).",
              },
              {
                "examples": [],
                "notes": [],
                "value": "A group of fish that migrate, or ascend a river for the purpose of spawning.",
              },
            ],
            "pos": "n",
          },
          "v": {
            "definitions": [
              {
                "definitions": [
                  {
                    "examples": [
                      {
                        "value": "The dog runs after the cat.",
                      },
                      {
                        "value": "The horse ran away.",
                      },
                    ],
                    "notes": [],
                    "value": "(vertebrates) To move swiftly (figuratively)",
                  },
                  {
                    "examples": [],
                    "notes": [],
                    "value": "(fluids) To flow.",
                  },
                  {
                    "examples": [],
                    "notes": [],
                    "value": "(nautical, of a vessel) To sail before the wind, in distinction from reaching or sailing close-hauled.",
                  },
                  {
                    "examples": [],
                    "notes": [],
                    "value": "(social) To carry out an activity.",
                  },
                  {
                    "examples": [],
                    "notes": [],
                    "value": "To extend or persist, statically or dynamically, through space or time.",
                  },
                  {
                    "examples": [],
                    "notes": [],
                    "value": "(transitive) To execute or carry out a plan, procedure or program.",
                  },
                ],
                "description": "A number of verb senses",
              },
            ],
            "pos": "v",
          },
        },
      },
    ],
    "term": "run",
  },
]
`;

exports[`Dictionary > can split terms properly 1`] = `
[
  {
    "etymologies": [
      {
        "description": "Latin root",
        "senses": {
          "n": {
            "definitions": [
              {
                "examples": [
                  {
                    "value": "There goes a cat!",
                  },
                ],
                "notes": [
                  {
                    "examples": [
                      {
                        "value": "Some example",
                      },
                    ],
                    "value": "Some definition note",
                  },
                ],
                "value": "a cat",
              },
            ],
            "pos": "n",
          },
        },
      },
    ],
    "term": "cat",
  },
  {
    "etymologies": [
      {
        "description": "Latin root",
        "pronunciation": "dooooog",
        "senses": {
          "un": {
            "definitions": [
              {
                "examples": [],
                "notes": [],
                "value": "a dog",
              },
            ],
            "pos": "un",
          },
        },
      },
    ],
    "term": "dog",
  },
]
`;

exports[`Dictionary > lookup > can split terms 1`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "definitions": [
                {
                  "examples": [
                    {
                      "value": "There goes a cat!",
                    },
                  ],
                  "notes": [
                    {
                      "examples": [
                        {
                          "value": "Some example",
                        },
                      ],
                      "value": "Some definition note",
                    },
                  ],
                  "value": "a cat",
                },
              ],
              "pos": "n",
            },
          },
        },
      ],
      "term": "cat",
    },
    {
      "etymologies": [
        {
          "description": "Latin root",
          "pronunciation": "dooooog",
          "senses": {
            "un": {
              "definitions": [
                {
                  "examples": [],
                  "notes": [],
                  "value": "a dog",
                },
              ],
              "pos": "un",
            },
          },
        },
      ],
      "term": "dog",
    },
  ],
]
`;

exports[`Dictionary > lookup > looks up terms properly 1`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "definitions": [
                {
                  "examples": [
                    {
                      "value": "There goes a cat!",
                    },
                  ],
                  "notes": [
                    {
                      "examples": [
                        {
                          "value": "Some example",
                        },
                      ],
                      "value": "Some definition note",
                    },
                  ],
                  "value": "a cat",
                },
              ],
              "pos": "n",
            },
          },
        },
      ],
      "term": "cat",
    },
  ],
]
`;
